src/engine.c

```diff
@@ -4205,29 +4230,42 @@ int settle_needed(game *g, int who, int which, int special[], int num_special,
 		/* Check for extra military */
 		if (o_ptr->code & P3_EXTRA_MILITARY)
 		{
+			/* Skip generic powers taken into account in total_military */
+			if (!(o_ptr->code & P3_CONDITIONAL_MILITARY)) continue;
+
 			/* Skip powers that require good consumption */
 			if (o_ptr->code & P3_CONSUME_RARE) continue;
 			if (o_ptr->code & P3_CONSUME_ALIEN) continue;
 
+			/* Skip powers that require discarding an active card */
+			if (o_ptr->code & P3_DISCARD) continue;
+
+			/* Check for specificity */
+
 			/* Check for specific good required */
-			if (((o_ptr->code & P3_NOVELTY) && good == GOOD_NOVELTY)||
-			    ((o_ptr->code & P3_RARE) && good == GOOD_RARE) ||
-			    ((o_ptr->code & P3_GENE) && good == GOOD_GENE) ||
-			    ((o_ptr->code & P3_ALIEN) && good == GOOD_ALIEN))
+			if (((o_ptr->code & P3_NOVELTY) && good != GOOD_NOVELTY)||
+			    ((o_ptr->code & P3_RARE) && good != GOOD_RARE) ||
+			    ((o_ptr->code & P3_GENE) && good != GOOD_GENE) ||
+			    ((o_ptr->code & P3_ALIEN) && good != GOOD_ALIEN))
 			{
-				/* Add value to military */
-				military += o_ptr->value;
+				/* Skip power */
 				continue;
 			}
 
 			/* Check for against rebels */
 			if ((o_ptr->code & P3_AGAINST_REBEL) &&
-			    (t_ptr->d_ptr->flags & FLAG_REBEL))
+			    !(t_ptr->d_ptr->flags & FLAG_REBEL))
 			{
-				/* Add value to military */
-				military += o_ptr->value;
+				/* Skip power */
 				continue;
 			}
+
+			/* Military power does not require payment and applies */
```

Comment:
At this point, we only consider extra-military that does no require payment, cf the code that skip extra-military that needs payment 35 lines above.

In settle_callback, a power requiring payment is processed only if selected by the player and passed through the "special" list argument.